File: src/main/java/nl/wernerkroneman/Drawy/ParseTreeMatcher/PhrasePatternBuilder.java
package nl.wernerkroneman.Drawy.ParseTreeMatcher;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

/**
 * Convenient PhraseTreePatternNode builder facilitating
 * an internal DSL to specify a PhraseTree pattern.
 */
public class PhrasePatternBuilder {
    private Pattern word;
    private Pattern nature;
    private Pattern role;
    private String name;
    private int repeat = 1;
    private List<PhrasePattern> children = null;

    public PhrasePatternBuilder setWord(Pattern word) {
        this.word = word;
        return this;
    }

    public PhrasePatternBuilder setNature(Pattern nature) {
        this.nature = nature;
        return this;
    }

    public PhrasePatternBuilder setRole(Pattern role) {
        this.role = role;
        return this;
    }


    public PhrasePatternBuilder setWord(String word) {
        this.word = Pattern.compile(word);
        return this;
    }

    public PhrasePatternBuilder setNature(String nature) {
        this.nature = Pattern.compile(nature);
        return this;
    }

    public PhrasePatternBuilder setRole(String role) {
        this.role = Pattern.compile(role);
        return this;
    }

    public PhrasePatternBuilder setName(String name) {
        this.name = name;
        return this;
    }

    public PhrasePatternBuilder setRepeat(int repeat) {
        this.repeat = repeat;
        return this;
    }

    public PhrasePatternBuilder setChildren(List<PhrasePattern> children) {
        this.children = children;
        return this;
    }

    public PhrasePatternBuilder setChildren(PhrasePattern... children) {
        this.children = Arrays.asList(children);
        return this;
    }

    public PhrasePatternBuilder addChild(PhrasePattern child) {
        if (this.children == null) {
            this.children = new ArrayList<>();
        }
        this.children.add(child);
        return this;
    }

    public PhrasePattern create() {
        return new PhrasePattern(word, nature, role, name, repeat, children);
    }

    public PhrasePatternBuilder setAnyChildren() {
        this.children = null;
        return this;
    }

    public PhrasePatternBuilder setNoChildren() {

        if (this.children == null) {
            this.children = new ArrayList<>();
        }

        this.children.clear();

        return this;
    }
}

File: src/main/java/nl/wernerkroneman/Drawy/ParseTreeMatcher/PhrasePatternBuilder.kt
/*
 * Copyright (c) 2017 Werner Kroneman
 *
 * This file is part of DrawyMcDrawface.
 *
 * DrawyMcDrawface is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DrawyMcDrawface is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DrawyMcDrawface.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.wernerkroneman.Drawy.ParseTreeMatcher

import java.util.*
import java.util.regex.Pattern
import java.util.regex.Pattern.CASE_INSENSITIVE

/**
 * Convenient PhraseTreePatternNode builder facilitating
 * an internal DSL to specify a PhraseTree pattern.
 */
class PhrasePatternBuilder {
    private var word: Pattern? = null
    private var nature: Pattern? = null
    private var role: Pattern? = null
    private var name: String? = null
    private var repeatMin: Int = 1
    private var repeatMax: Int? = 1
    private var children: MutableList<PhrasePattern>? = null

    fun word(word: Pattern): PhrasePatternBuilder {
        this.word = word
        return this
    }

    fun nature(nature: Pattern): PhrasePatternBuilder {
        this.nature = nature
        return this
    }

    fun role(role: Pattern): PhrasePatternBuilder {
        this.role = role
        return this
    }


    fun word(word: String): PhrasePatternBuilder {
        this.word = Pattern.compile(word, CASE_INSENSITIVE)
        return this
    }

    fun nature(nature: String): PhrasePatternBuilder {
        this.nature = Pattern.compile(nature)
        return this
    }

    fun role(role: String): PhrasePatternBuilder {
        this.role = Pattern.compile(role)
        return this
    }

    fun name(name: String): PhrasePatternBuilder {
        this.name = name
        return this
    }

    fun children(children: MutableList<PhrasePattern>): PhrasePatternBuilder {
        this.children = children
        return this
    }

    fun children(vararg children: PhrasePattern): PhrasePatternBuilder {
        this.children = Arrays.asList(*children)
        return this
    }

    fun child(child: PhrasePattern): PhrasePatternBuilder {
        if (this.children == null) {
            this.children = ArrayList<PhrasePattern>()
        }
        this.children!!.add(child)
        return this
    }

    fun child(build: PhrasePatternBuilder.()->Unit): PhrasePatternBuilder {
        return child(buildPattern(build))
    }

    fun create(): PhrasePattern {
        return PhrasePattern(word, nature, role, name, repeatMin, repeatMax, children)
    }

    fun anyChildren(): PhrasePatternBuilder {
        this.children = null
        return this
    }

    fun noChildren(): PhrasePatternBuilder {

        if (this.children == null) {
            this.children = ArrayList<PhrasePattern>()
        }

        this.children!!.clear()

        return this
    }

    fun repeat(repeatMin: Int, repeatMax: Int? = repeatMin): PhrasePatternBuilder {
        this.repeatMin = repeatMin
        this.repeatMax = repeatMax
        return this
    }

    fun repeatAny(): PhrasePatternBuilder {
        repeatMin = 1
        repeatMax = null
        return this
    }

    fun optional(optional: Boolean = true): PhrasePatternBuilder {
        repeatMin = 0
        repeatMax = 1
        return this
    }
}

fun buildPattern(build: PhrasePatternBuilder.()->Unit) : PhrasePattern {
    val builder = PhrasePatternBuilder()
    builder.build()
    return builder.create()
}